### ISSUE TITLE
0.59.5 - Evohome TRV re-prompting for already discovered devices

### ISSUE THREAD

**Issue by peternash** [ramses-rf/ramses_cc#954]
I'm running current masters of both ramses_rf and ramses_cc (0.59.5+) on my test system .  These versions appear to fix [#931](https://github.com/ramses-rf/ramses_cc/issues/931),  [#947](https://github.com/ramses-rf/ramses_cc/issues/947) and [#948](https://github.com/ramses-rf/ramses_cc/issues/948).

However, I've noticed that the discovery process can re-prompt for devices that have already been discovered, specifically it can re-prompt for a TRV _without_ zone information even when that TRV has already been discovered _with_ it's zone.  Accepting this prompt will result in the TRV becoming a `heat_orphan` and then later being moved back into the correct zone.

I've also noticed that I keep seeing prompts for my "foreign" second HGI even though it's been marked as "not_mine" in the schema.

Eventually discovery "settles down" to a stable (and correct) schema but a restart of HA starts the process again.

Attached is an example of a restart of HA after full schema discovery:

Schema:

[Schema.txt](https://github.com/user-attachments/files/31058084/Schema.txt)

Logs:

[home-assistant.log](https://github.com/user-attachments/files/31058102/home-assistant.log)
[packet_log.log](https://github.com/user-attachments/files/31058101/packet_log.log)

"Discovery" prompt:

<img width="592" height="696" alt="Image" src="https://github.com/user-attachments/assets/b1abb209-e28e-4950-9732-6c19c4384628" />

**Comment by wimpie70** [comment]
That's annoying...I'll have a look

**Comment by peternash** [comment]
Not sure if it's related but I'm also seeing this periodically in the log:
```
2026-08-14 13:16:36.662 INFO (MainThread) [custom_components.ramses_cc.coordinator] sync_learned_topology: removed_devices=set()
2026-08-14 13:16:36.668 DEBUG (MainThread) [custom_components.ramses_cc.coordinator] sync_learned_topology: enriched=None
2026-08-14 13:16:36.668 INFO (MainThread) [custom_components.ramses_cc.coordinator] No topology changes, but device_comments refreshed from scan engine — persisting updated comments
2026-08-14 13:16:36.668 DEBUG (MainThread) [custom_components.ramses_cc] Config entry 01KZ181MJNMJN4B7DAF2K5P7A6 updated, but reload suppressed (accept flow)
2026-08-14 13:16:36.671 INFO (MainThread) [custom_components.ramses_cc.coordinator] Saving state: discovery_manager=True, cached=False, discovery_devices=0
2026-08-14 13:16:36.689 ERROR (MainThread) [homeassistant.helpers.storage] Error writing config for ramses_cc: Bad data at $.data.client_state.packets.2026-08-14T13:11:35.751000+01:00.payload.msg_type=b'\x00\x10'(<class 'bytes'>, $.data.client_state.packets.2026-08-14T13:11:35.751000+01:00.payload.payload_data=b'\x01\xa0\x00/O\xaav0.59.5'(<class 'bytes'>
```

**Comment by wimpie70** [comment]
I've a report on this. I'll come up with a fix once I have ha_sim_test stable again after the recent PR's.

# Issue 954 analysis — re-prompting + packet serialization failure

Target: https://github.com/ramses-rf/ramses_cc/issues/954
Reporter: @peternash
Status: open, two comments

---

## Summary

Issue 954 reports that after HA restart, discovery re-prompts for
devices that have already been discovered and placed in the schema.
Two specific symptoms:

1. A TRV already placed in a zone is re-prompted **without** zone info.
   Accepting it makes it a `heat_orphan`, then it gets moved back to the
   correct zone later.
2. A foreign second HGI (marked `not_mine`) is re-prompted every cycle.

A follow-up comment adds a third symptom: a packet serialization error
that aborts the entire save cycle, causing discovery state loss on
restart.

---

## Bug A — Foreign HGI (18:) re-prompted every cycle

### Root cause

`get_devices()` in `custom_components/ramses_cc/discovery.py` (line
1185-1191) merges scan-engine devices with metadata, defaulting any
device without metadata to `DeviceMetadata()` which has
`status=NEW`:

```python
all_ids = set(engine_devices.keys()) | set(self._metadata.keys())
for device_id in all_ids:
    meta = self._metadata.get(device_id, DeviceMetadata())  # defaults to NEW
```

Both `sync_with_schema()` (line 493) and `check_for_new_devices()`
(line 1775) **explicitly skip 18: devices** — they never create
metadata for them. But `get_devices()` does **not** skip 18: devices.
So a foreign HGI that the scan engine sees on the RF network:

1. `sync_with_schema` skips it → no metadata created
2. `check_for_new_devices` skips it → no metadata created
3. `get_devices(status=NEW)` includes it because the default
   `DeviceMetadata()` has `status=NEW`
4. The `review_discovered` form shows it every time

This persists across restarts because the HGI is never given ACCEPTED
metadata.

### Fix

`get_devices()` must skip 18: devices (or default them to a
non-NEW status). The skip already exists in `sync_with_schema` and
`check_for_new_devices` — `get_devices` needs the same guard:

```python
for device_id in all_ids:
    if device_id.startswith("18:"):
        continue  # HGI gateways are not discoverable
    meta = self._metadata.get(device_id)
    if meta is None:
        if device_id in self._schema_device_ids:
            continue  # already in schema, not a new discovery
        meta = DeviceMetadata()  # genuinely new
```

### Code reference

- `discovery.py` lines 1185-1191 (`get_devices` default to NEW)
- `discovery.py` line 493 (`sync_with_schema` skips 18:)
- `discovery.py` line 1775 (`check_for_new_devices` skips 18:)

---

## Bug B — TRV re-prompted without zone (extraction path discrepancy)

### Root cause

There are **two different device-ID extraction paths** that produce
different sets:

- **Save path** (`_extract_schema_device_ids` →
  `_derive_known_list_from_schema` in `coordinator.py` line 1031-1034):
  **includes** global `orphans_heat` / `orphans_hvac` device IDs.
- **Startup path** (`_extract_device_ids_from_stripped` in
  `coordinator.py` line 901-907): **skips** `SZ_ORPHANS_HEAT` and
  `SZ_ORPHANS_HVAC` as top-level keys.

This means devices that are only in global `orphans_heat` are:

1. Included in `_discovery_filter_ids` during save → metadata is saved
2. **NOT included** in `schema_device_ids` during startup →
   `sync_with_schema` marks them REMOVED
3. `check_for_new_devices` re-marks them NEW → re-prompted

For a TRV in a zone, both paths should find it from
`zones[].actuators`. But if the TRV's metadata was lost (e.g.
.storage timing issue) or the save filter excluded it,
`get_devices` defaults to NEW regardless of schema placement.

### Fix

`_extract_device_ids_from_stripped` should include global
`orphans_heat` / `orphans_hvac` device IDs, matching
`_derive_known_list_from_schema`. The skip at line 901-907 should
extract device IDs from the orphan lists, not skip them entirely:

```python
if key in (SZ_MAIN_TCS, "transport_constructor"):
    continue
if key in (SZ_ORPHANS_HEAT, SZ_ORPHANS_HVAC):
    if isinstance(value, list):
        device_ids.update(value)
    continue
```

### Code reference

- `coordinator.py` lines 882-947 (`_extract_device_ids_from_stripped`,
  skips global orphans)
- `coordinator.py` lines 1031-1034
  (`_derive_known_list_from_schema`, includes global orphans)

---

## Bug C — Packet serialization failure (save cycle aborts)

### Root cause

`get_state()` in `ramses_rf/gateway.py` (line 377) passes the **raw
payload dataclass** into the state dict:

```python
state_dict[dtm_str] = {
    ...
    "payload": msg.payload,  # raw Phase 6 dataclass
    ...
}
```

`PuzzlePayload` (7FFF) in `ramses_rf/payloads/system.py` (lines
1781-1782) has `bytes` fields:

```python
msg_type: bytes
payload_data: bytes
```

Unlike every other payload class, `PuzzlePayload` has **no `to_dict()`
method**. When HA's storage layer tries to JSON-serialize the state
dict, it rejects `bytes` objects:

```
Error writing config for ramses_cc: Bad data at
$.data.client_state.packets.*.payload.msg_type=b'\x00\x10'(<class 'bytes'>
```

This causes the **entire save cycle to fail** — not just the
problematic packet, but all state (schema, packets, discovery, HVAC)
is lost. This is why the user sees re-prompting after every restart:
the discovery state is never persisted because the packet
serialization error aborts the save.

### Also at risk

`OpenThermMsgPayload` in `ramses_rf/payloads/opentherm.py` has
`raw_value: bytes`. It does have a `to_dict()` that converts
`raw_value` to `.hex().upper()` (line 124), but `get_state()` passes
the raw object, not `to_dict()`, so it's also vulnerable if any code
path stores the raw dataclass.

### Fix (two parts)

**Part 1 — ramses_rf: add `to_dict()` to `PuzzlePayload`:**

```python
def to_dict(self, msg: Any = None) -> dict[str, Any]:
    """Convert puzzle diagnostic payload to legacy dictionary format."""
    return {
        "msg_type": self.msg_type.hex(),
        "payload_data": self.payload_data.hex(),
    }
```

**Part 2 — ramses_rf: `get_state()` should call `to_dict()` on
payloads:**

```python
payload = msg.payload
if hasattr(payload, "to_dict"):
    try:
        payload = payload.to_dict(msg) if ... else payload.to_dict()
    except Exception:
        payload = str(payload)
state_dict[dtm_str] = {
    ...
    "payload": payload,
    ...
}
```

This protects against future payload classes that forget `to_dict()`
and against any `bytes` fields that slip through.

### Code reference

- `ramses_rf/gateway.py` line 377 (`get_state` passes raw payload)
- `ramses_rf/payloads/system.py` lines 1759-1805 (`PuzzlePayload`,
  no `to_dict()`)
- `ramses_rf/payloads/opentherm.py` lines 44-45
  (`OpenThermMsgPayload`, `raw_value: bytes`)

---

## Connection to ha_sim_test failures

| Recipe | Symptom | Likely cause |
|--------|---------|--------------|
| R26 | 04:200099 not in schema after profile load | Bug B — device in global `orphans_heat` after stripping but not in `schema_device_ids` |
| R50 | "no scan_state in .storage" | Bug C — save cycle fails, `scan_state` never persisted |
| R30 | FAN class=None, _bound=None | Possibly Bug C — save failure prevents schema persistence |
| R19/R22 | THM comment empty | Possibly Bug B — device not in `schema_device_ids`, metadata lost |

Bug C is the most urgent — it causes **data loss** (entire save cycle
fails), which explains why the user sees re-prompting after every
restart.

---

## Suggested fix order

1. **Bug C** (ramses_rf) — add `to_dict()` to `PuzzlePayload`, make
   `get_state()` call `to_dict()`. This unblocks the save cycle and
   may resolve several ha_sim_test failures (R50, R30).
2. **Bug A** (ramses_cc) — skip 18: in `get_devices()`. Small,
   uncontroversial.
3. **Bug B** (ramses_cc) — align the two extraction paths. Small but
   needs careful testing to ensure no devices are lost from the
   known_list.

---

## Numbered points for future comments

- **Point A** — Foreign HGI re-prompted: `get_devices()` defaults 18:
  devices to NEW; `sync_with_schema` and `check_for_new_devices` skip
  them but `get_devices` doesn't.
- **Point B** — TRV re-prompted without zone:
  `_extract_device_ids_from_stripped` skips global `orphans_heat`/
  `orphans_hvac` but `_derive_known_list_from_schema` includes them.
- **Point C** — Packet serialization failure: `PuzzlePayload` (7FFF)
  has `bytes` fields and no `to_dict()`; `get_state()` passes the raw
  dataclass; HA storage rejects `bytes`; entire save cycle aborts.
- **Point D** — Bug C is the most urgent: it causes data loss, which
  is the root cause of the restart re-prompting (discovery state is
  never persisted).

**Comment by wimpie70** [comment]
Actually, the recipes mentioned here are what I was  working on right now...

**Comment by silverailscolo** [comment]
@peternash Fixes are in 0.59.6, just out in HACS. Please report whether that fixes it for you.

**Comment by peternash** [comment]
@silverailscolo - I'm testing with 0.59.6 and the problems raised in this issue have been fixed.  I'm seeing a slightly different issue - neither the local of foreign HGI is being added to the schema correctly but I'll raise that as a new issue.
